tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -674,6 +688,689 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+		relids = IS_OTHER_REL(joinrel) ? joinrel->top_parent_relids : joinrel->relids;
+
+		if (bms_is_subset(phinfo->ph_eval_at, relids) &&
+			bms_nonempty_difference(relids, phinfo->ph_eval_at))
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg(
+						 "join pushdown failed, placeholder is not suitable to be pushed down")));
+	}
+
+	fpinfo->outerrel = outerrel;
+	fpinfo->innerrel = innerrel;
+	fpinfo->jointype = jointype;
+
+	/*
+	 * By default, both the input relations are not required to be deparsed as
+	 * subqueries, but there might be some relations covered by the input
+	 * relations that are required to be deparsed as subqueries, so save the
+	 * relids of those relations for later use by the deparser.
+	 */
+	fpinfo->make_outerrel_subquery = false;
+	fpinfo->make_innerrel_subquery = false;
+	Assert(bms_is_subset(fpinfo_o->lower_subquery_rels, outerrel->relids));
+	Assert(bms_is_subset(fpinfo_i->lower_subquery_rels, innerrel->relids));
+	fpinfo->lower_subquery_rels =
+		bms_union(fpinfo_o->lower_subquery_rels, fpinfo_i->lower_subquery_rels);
+
+	/*
+	 * Pull the other remote conditions from the joining relations into join
+	 * clauses or other remote clauses (remote_conds) of this relation
+	 * wherever possible. This avoids building subqueries at every join step.
+	 *
+	 * For an inner join, clauses from both the relations are added to the
+	 * other remote clauses. For LEFT and RIGHT OUTER join, the clauses from
+	 * the outer side are added to remote_conds since those can be evaluated
+	 * after the join is evaluated. The clauses from inner side are added to
+	 * the joinclauses, since they need to be evaluated while constructing the
+	 * join.
+	 *
+	 * For a FULL OUTER JOIN, the other clauses from either relation can not
+	 * be added to the joinclauses or remote_conds, since each relation acts
+	 * as an outer relation for the other.
+	 *
+	 * The joining sides can not have local conditions, thus no need to test
+	 * shippability of the clauses being pulled up.
+	 */
+	switch (jointype)
+	{
+		case JOIN_INNER:
+#if PG14_GE
+			fpinfo->remote_conds = list_concat(fpinfo->remote_conds, fpinfo_i->remote_conds);
+			fpinfo->remote_conds = list_concat(fpinfo->remote_conds, fpinfo_o->remote_conds);
+#else
+			fpinfo->remote_conds =
+				list_concat(fpinfo->remote_conds, list_copy(fpinfo_i->remote_conds));
+			fpinfo->remote_conds =
+				list_concat(fpinfo->remote_conds, list_copy(fpinfo_o->remote_conds));
+#endif
+			break;
+
+		case JOIN_LEFT:
+#if PG14_GE
+			fpinfo->joinclauses = list_concat(fpinfo->joinclauses, fpinfo_i->remote_conds);
+			fpinfo->remote_conds = list_concat(fpinfo->remote_conds, fpinfo_o->remote_conds);
+#else
+			fpinfo->joinclauses =
+				list_concat(fpinfo->joinclauses, list_copy(fpinfo_i->remote_conds));
+			fpinfo->remote_conds =
+				list_concat(fpinfo->remote_conds, list_copy(fpinfo_o->remote_conds));
+#endif
+			break;
+
+		case JOIN_RIGHT:
+#if PG14_GE
```

Comment:
Looks like we need a test case for JOIN_RIGHT

Reply:
Regarding the test cases / test coverage: The code is from PG upstream and is not triggered in the current implementation. We could remove these conditions and the support for other join types. However, if we decide to support these joins, it makes a lot of work to determine which statements have been removed (especially if these are single statements and not obvious things like entire functions or parameters).

Reply:
Can we make dead code clear,  by wrapping it in `#if ENABLE_DEAD_CODE` or marking it with `pg_unreachable()`, or similar so that the code is there but you can avoid having code coverage complain and it is obvious to the reader that the code isn't executed.